File: AgoraChat/Code/CallKit/Conference/Invite/ConfInviteUsersViewController.m
//
//  ConfInviteUsersViewController.m
//  ChatDemo-UI3.0
//
//  Created by XieYajie on 2018/9/20.
//  Copyright © 2018 XieYajie. All rights reserved.
//

#import "ConfInviteUsersViewController.h"

#import "EMRealtimeSearch.h"
#import "ConfInviteUserCell.h"
#import "UserInfoStore.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import "MJRefresh.h"

@interface ConfInviteUsersViewController ()<UITableViewDataSource, UITableViewDelegate, UISearchBarDelegate>

@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UISearchBar *searchBar;
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) UITableView *searchTableView;

@property (nonatomic, strong) NSArray *excludeUsers;
@property (nonatomic, strong) NSString *groupId;

@property (nonatomic, strong) NSString *cursor;
@property (nonatomic) BOOL isSearching;
@property (nonatomic, strong) NSMutableArray *dataArray;
@property (nonatomic, strong) NSMutableArray *searchDataArray;
@property (nonatomic, strong) NSMutableArray *inviteUsers;

@end

@implementation ConfInviteUsersViewController

- (instancetype)initWithGroupId:(NSString *)groupId excludeUsers:(NSArray *)excludeUserList {
    if (self = [super init]) {
        _groupId = groupId;
        _excludeUsers = excludeUserList;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refreshTableView) name:USERINFO_UPDATE object:nil];
    // Do any additional setup after loading the view.
    _dataArray = [NSMutableArray array];
    _searchDataArray = [NSMutableArray array];
    self.inviteUsers = [NSMutableArray array];
    
    [self _setupSubviews];
    [self tableViewDidTriggerHeaderRefresh];
}

- (void)dealloc
{
    self.searchBar.delegate = nil;
}

#pragma mark - Subviews

- (void)_setupSubviews
{
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.textColor = [UIColor blackColor];
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.titleLabel.font = [UIFont systemFontOfSize:18];
    self.titleLabel.text = NSLocalizedString(@"title.selectMembers", nil);
    
    [self.view addSubview:self.titleLabel];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.view).offset(45);
        make.right.equalTo(self.view).offset(-45);
        make.top.equalTo(self.view).offset(13);
        make.height.equalTo(@25);
    }];
    
    UIButton *confirmButton = [[UIButton alloc] init];
    confirmButton.titleLabel.font = [UIFont systemFontOfSize:15];
    [confirmButton setTitle:NSLocalizedString(@"close", nil) forState:UIControlStateNormal];
    [confirmButton setTitleColor:[UIColor colorWithRed:8 / 255.0 green:115 / 255.0 blue:222 / 255.0 alpha:1.0] forState:UIControlStateNormal];
    [confirmButton setTitleColor:[UIColor grayColor] forState:UIControlStateHighlighted];
    [confirmButton addTarget:self action:@selector(confirmAction) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:confirmButton];
    [confirmButton mas_makeConstraints:^(MASConstraintMaker *make) {
        //make.left.equalTo(self.titleLabel.mas_right);
        make.right.equalTo(self.view).offset(-10);
        make.top.equalTo(self.titleLabel);
        make.bottom.equalTo(self.titleLabel);
    }];
    
    self.searchBar = [[UISearchBar alloc] init];
    self.searchBar.delegate = self;
    self.searchBar.barTintColor = [UIColor whiteColor];
    self.searchBar.searchBarStyle = UISearchBarStyleMinimal;
    UITextField *searchField = [self.searchBar valueForKey:@"searchField"];
    CGFloat color = 245 / 255.0;
    searchField.backgroundColor = [UIColor colorWithRed:color green:color blue:color alpha:1.0];
    self.searchBar.placeholder = NSLocalizedString(@"serchContact", nil);
    [self.view addSubview:self.searchBar];
    [self.view sendSubviewToBack:self.searchBar];
    [self.searchBar mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.titleLabel.mas_bottom).offset(12);
        make.left.equalTo(self.view).offset(12);
        make.right.equalTo(self.view).offset(-12);
    }];
    
    self.searchTableView = [[UITableView alloc] init];
    self.searchTableView.delegate = self;
    self.searchTableView.dataSource = self;
    self.searchTableView.rowHeight = 54;
    
    _tableView = [[UITableView alloc] init];
    _tableView.dataSource = self;
    _tableView.delegate = self;
    _tableView.rowHeight = 54;
    [self.view addSubview:_tableView];
    [_tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.searchBar.mas_bottom);
        make.left.right.bottom.equalTo(self.view);
    }];
    
    UINib *nib = [UINib nibWithNibName:@"ConfInviteUserCell" bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:@"ConfInviteUserCell"];
    [self.searchTableView registerNib:nib forCellReuseIdentifier:@"ConfInviteUserCell"];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.isSearching ? [self.searchDataArray count] : [self.dataArray count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"ConfInviteUserCell";
    ConfInviteUserCell *cell = (ConfInviteUserCell *)[tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    
    NSString *username = self.isSearching ? [self.searchDataArray objectAtIndex:indexPath.row] : [self.dataArray objectAtIndex:indexPath.row];
    cell.nameLabel.text = username;
    cell.imgView.image = [UIImage imageNamed:@"defaultAvatar"];
    cell.isChecked = [self.inviteUsers containsObject:username];
    AgoraChatUserInfo *userInfo = [UserInfoStore.sharedInstance getUserInfoById:username];
    if (userInfo) {
        if (userInfo.nickname.length > 0) {
            cell.nameLabel.text = userInfo.nickname;
        }
        if (userInfo.avatarUrl.length > 0) {
            NSURL *url = [NSURL URLWithString:userInfo.avatarUrl];
            if (url) {
                [cell.imgView sd_setImageWithURL:url];
            }
        }
    } else {
        [UserInfoStore.sharedInstance fetchUserInfosFromServer:@[username]];
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    NSString *username = self.isSearching ? [self.searchDataArray objectAtIndex:indexPath.row] : [self.dataArray objectAtIndex:indexPath.row];
    ConfInviteUserCell *cell = (ConfInviteUserCell *)[tableView cellForRowAtIndexPath:indexPath];
    BOOL isChecked = [self.inviteUsers containsObject:username];
    if (isChecked) {
        [self.inviteUsers removeObject:username];
    } else {
        [self.inviteUsers addObject:username];
    }
    cell.isChecked = !isChecked;
    
    NSUInteger count = self.inviteUsers.count;
    self.titleLabel.text = count == 0 ? NSLocalizedString(@"title.selectMembers", nil) : [NSString stringWithFormat:@"%@(%lu)", NSLocalizedString(@"title.selectMembers", nil), (unsigned long)count];
}

#pragma mark - UISearchBarDelegate

- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar
{
    [searchBar setShowsCancelButton:YES];
    return YES;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    if (!self.isSearching) {
        self.isSearching = YES;
        [self.view addSubview:self.searchTableView];
        [self.searchTableView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self.tableView);
            make.left.equalTo(self.tableView);
            make.right.equalTo(self.tableView);
            make.bottom.equalTo(self.tableView);
        }];
    }
    
    __weak typeof(self) weakSelf = self;
    [EMRealtimeSearch.shared realtimeSearchWithSource:self.dataArray searchText:searchBar.text collationStringSelector:nil resultBlock:^(NSArray *results) {
        if ([results count] > 0) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.searchDataArray removeAllObjects];
                [weakSelf.searchDataArray addObjectsFromArray:results];
                [self.searchTableView reloadData];
            });
        }
    }];
}

- (BOOL)searchBar:(UISearchBar *)searchBar shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    if ([text isEqualToString:@"\n"]) {
        [searchBar resignFirstResponder];
        return NO;
    }
    return YES;
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [[EMRealtimeSearch shared] realtimeSearchStop];
    [searchBar setShowsCancelButton:NO];
    [searchBar resignFirstResponder];

    self.isSearching = NO;
    [self.searchDataArray removeAllObjects];
    [self.searchTableView removeFromSuperview];
    [self.searchTableView reloadData];
    [self.tableView reloadData];
}

- (void)refreshTableView
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if(self.view.window)
            [self.tableView reloadData];
    });
}

#pragma mark - Data

- (NSArray *)_getInvitableUsers:(NSArray *)aAllUsers
{
    NSMutableArray *retNames = [[NSMutableArray alloc] init];
    [retNames addObjectsFromArray:aAllUsers];
    
    NSString *loginName = [AgoraChatClient.sharedClient.currentUsername lowercaseString];
    if ([retNames containsObject:loginName]) {
        [retNames removeObject:loginName];
    }
    
    for (NSString *name in self.excludeUsers) {
        if ([retNames containsObject:name]) {
            [retNames removeObject:name];
        }
    }
    
    return retNames;
}

- (void)_fetchGroupMembersWithIsHeader:(BOOL)aIsHeader
{
    NSInteger pageSize = 50;
    __weak typeof(self) weakSelf = self;
    [self showHudInView:self.view hint:NSLocalizedString(@"fetchingGroupMember...", nil)];
    [AgoraChatClient.sharedClient.groupManager getGroupMemberListFromServerWithId:self.groupId cursor:self.cursor pageSize:pageSize completion:^(AgoraChatCursorResult *aResult, AgoraChatError *aError) {
        if (aError) {
            [weakSelf hideHud];
            [weakSelf.tableView.mj_header endRefreshing];
            [weakSelf.tableView.mj_footer endRefreshing];
            
            [weakSelf showHint:[[NSString alloc] initWithFormat:NSLocalizedString(@"fetchGroupMemberFail", nil), aError.errorDescription]];
            return ;
        }
        
        weakSelf.cursor = aResult.cursor;
        
        if (aIsHeader) {
            [weakSelf.dataArray removeAllObjects];
            
            AgoraChatError *error = nil;
            AgoraChatGroup *group = [AgoraChatClient.sharedClient.groupManager getGroupSpecificationFromServerWithId:weakSelf.groupId error:&error];
            if (!error) {
                NSArray *owners = [weakSelf _getInvitableUsers:@[group.owner]];
                [weakSelf.dataArray addObjectsFromArray:owners];
                
                NSArray *admins = [weakSelf _getInvitableUsers:group.adminList];
                [weakSelf.dataArray addObjectsFromArray:admins];
            }
        }
        
        [weakSelf hideHud];
        [weakSelf.tableView.mj_header endRefreshing];
        [weakSelf.tableView.mj_footer endRefreshing];
        
        NSArray *usernames = [weakSelf _getInvitableUsers:aResult.list];
        [weakSelf.dataArray addObjectsFromArray:usernames];
        [weakSelf.tableView reloadData];
    }];
}

- (void)tableViewDidTriggerHeaderRefresh
{
    self.cursor = @"";
    [self _fetchGroupMembersWithIsHeader:YES];
}

- (void)tableViewDidTriggerFooterRefresh
{
    [self _fetchGroupMembersWithIsHeader:NO];
}

#pragma mark - Action

- (void)confirmAction
{
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:YES completion:^{
        if (weakSelf.didSelectedUserList) {
            weakSelf.didSelectedUserList(self.inviteUsers);
        }
    }];
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/CallKit/Conference/Invite/ConfInviteUsersViewController.swift
//
//  ConfInviteUsersViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/18.
//

import UIKit

@objcMembers
class ConfInviteUsersViewController: UIViewController {

    @IBOutlet private weak var titleLabel: UILabel!
    @IBOutlet private weak var collectionView: UICollectionView!
    @IBOutlet private weak var tableView: UITableView!
    @IBOutlet private weak var searchTableView: UITableView!
    
    private let excludeUsers: [String]
    private let groupId: String
    private var inGroupArray: [String]
    private var userModelMap: [String: AgoraUserModel] = [:]
    private var dataArray: [String] = []
    private var searchDataArray: [String] = []
    private var inviteUsers: [String] = []
    private var isSearching = false
    private var cursor: String?
    
    private let realtimeSearchUtil = AgoraRealtimeSearchUtils()
    
    var didSelectedUserListHandle: ((_ users: [String]) -> Void)?
    
    init(groupId: String, excludeUsers: [String]) {
        self.groupId = groupId
        self.excludeUsers = excludeUsers
        self.inGroupArray = Array(excludeUsers)
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        NotificationCenter.default.addObserver(self, selector: #selector(refreshTableView), name: UserInfoDidChangeNotification, object: nil)
     
        self.collectionView.register(UINib(nibName: "ACDMemberCollectionCell", bundle: nil), forCellWithReuseIdentifier: "cell")
        
        let nib = UINib(nibName: "AgoraGroupMemberCell", bundle: nil)
        self.tableView.register(nib, forCellReuseIdentifier: "cell")
        self.searchTableView.register(nib, forCellReuseIdentifier: "cell")
        self.fetchGroupMembers(refresh: true)
    }

    @IBAction func confirmAction() {
        self.didSelectedUserListHandle?(self.inviteUsers)
        self.dismiss(animated: true)
    }
    
    @objc private func refreshTableView() {
        self.tableView.reloadData()
    }
    
    private func fetchGroupMembers(refresh: Bool) {
        let pageSize = 20
        MBProgressHUD.showAdded(to: self.view, animated: true)
        AgoraChatClient.shared().groupManager?.getGroupMemberListFromServer(withId: self.groupId, cursor: self.cursor, pageSize: pageSize, completion: { result, error in
            MBProgressHUD.hide(for: self.view, animated: true)
            self.tableView.mj_header?.endRefreshing()
            self.tableView.mj_footer?.endRefreshing()
            if let error = error {
                self.showHint(error.errorDescription)
                return
            }
            self.cursor = result?.cursor
            var allList: [String] = []
            if refresh {
                self.dataArray.removeAll()
                if let group = AgoraChatClient.shared().groupManager?.getGroupSpecificationFromServer(withId: self.groupId, error: nil) {
                    if let owner = group.owner {
                        allList.append(owner)
                    }
                    if let admin = group.adminList {
                        allList.append(contentsOf: admin)
                    }
                }
                self.loadUserInfo(userIds: self.dataArray)
            }
            if let list = result?.list as? [String] {
                allList.append(contentsOf: list)
            }
            self.loadUserInfo(userIds: allList)
            let resultList = self.getInvitableUsers(users: allList)
            self.dataArray.append(contentsOf: resultList)
            self.tableView.reloadData()
        })
    }
    
    private func loadUserInfo(userIds: [String]) {
        UserInfoStore.shared.fetchUserInfosFromServer(userIds: userIds) {
            for userId in userIds {
                let model = AgoraUserModel(hyphenateId: userId)
                if let userInfo = UserInfoStore.shared.getUserInfo(userId: userId) {
                    model.setUserInfo(userInfo)
                }
                self.userModelMap[userId] = model
            }
            self.tableView.reloadData()
            self.collectionView.reloadData()
        }
    }

    private func getInvitableUsers(users: [String]) -> [String] {
        var retNames = Array(users)
        if let loginName = AgoraChatClient.shared().currentUsername?.lowercased() {
            for i in 0..<retNames.count where retNames[i] == loginName {
                retNames.remove(at: i)
                break
            }
        }
        for name in self.excludeUsers {
            for i in 0..<retNames.count where retNames[i] == name {
                retNames.remove(at: i)
                break
            }
        }
        return retNames
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

extension ConfInviteUsersViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearching ? self.searchDataArray.count : self.dataArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? AgoraGroupMemberCell {
            let username = (self.isSearching ? self.searchDataArray : self.dataArray)[indexPath.row]
//            if let item = self.userModelMap[username] {
//                cell.nameLabel.text = item.showName
//                cell.iconImageView.setImage(withUrl: item.avatarURLPath, placeholder: item.defaultAvatar)
//            } else {
//                cell.nameLabel.text = username
//                cell.iconImageView.image = UIImage(named: "defaultAvatar")
//            }
            cell.isSelect = self.inviteUsers.contains(username)
            cell.model = self.userModelMap[username]
//            cell.selectImageView.image =
//            cell.isChecked = self.inviteUsers.contains(username)
        }
        return cell
    }
}

extension ConfInviteUsersViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let username = (self.isSearching ? self.searchDataArray : self.dataArray)[indexPath.row]
        let cell = tableView.cellForRow(at: indexPath) as? AgoraGroupMemberCell
        if self.inviteUsers.count + self.excludeUsers.count >= 3, cell?.isSelect != true {
            self.showHint("There can only be 3 people in the channel".localized)
            return
        }
        let isChecked = self.inviteUsers.contains(username)
        if isChecked {
            for i in 0..<self.inviteUsers.count where self.inviteUsers[i] == username {
                self.inviteUsers.remove(at: i)
                break
            }
            for i in 0..<self.inGroupArray.count where self.inGroupArray[i] == username {
                self.inGroupArray.remove(at: i)
                break
            }
        } else {
            self.inviteUsers.append(username)
            self.inGroupArray.append(username)
        }
        cell?.isSelect = !isChecked
        let count = self.inviteUsers.count
        let title = "Select Members".localized
        self.titleLabel.text = count == 0 ? title : "\(title)(\(count))"
        self.collectionView.reloadData()
    }
}

extension ConfInviteUsersViewController: UISearchBarDelegate {
    
    func searchBarShouldBeginEditing(_ searchBar: UISearchBar) -> Bool {
        searchBar.setShowsCancelButton(true, animated: true)
        return true
    }

    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        if !self.isSearching {
            self.isSearching = true
            self.searchTableView.isHidden = false
        }
        self.realtimeSearchUtil.realSearch(source: self.dataArray, keyword: searchBar.text ?? "") { result in
            self.searchDataArray.removeAll()
            if let result = result as? [String] {
                self.searchDataArray.append(contentsOf: result)
            }
            self.searchTableView.reloadData()
        }
    }
    
    func searchBar(_ searchBar: UISearchBar, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            searchBar.resignFirstResponder()
            return false
        }
        return true
    }

    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        self.realtimeSearchUtil.cancel()
        searchBar.setShowsCancelButton(false, animated: true)
        searchBar.resignFirstResponder()
        
        self.isSearching = false
        self.searchDataArray.removeAll()
        self.searchTableView.isHidden = true
        self.searchTableView.reloadData()
        self.tableView.reloadData()
    }
}

extension ConfInviteUsersViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.inGroupArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDMemberCollectionCell {
            let username = self.inGroupArray[indexPath.item]
            cell.username = username
            cell.model = self.userModelMap[username]
            cell.deleteEnable = !self.excludeUsers.contains(self.inGroupArray[indexPath.item])
        }
        return cell
    }
}

extension ConfInviteUsersViewController: UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let username = self.inGroupArray[indexPath.item]
        if excludeUsers.contains(username) {
            return
        }
        for i in 0..<self.inviteUsers.count where self.inviteUsers[i] == username {
            self.inviteUsers.remove(at: i)
            break
        }
        for i in 0..<self.inGroupArray.count where self.inGroupArray[i] == username {
            self.inGroupArray.remove(at: i)
            break
        }
        self.tableView.reloadData()
        self.collectionView.reloadData()
    }
}
